### DISCUSSION TITLE
Manual installation

### DISCUSSION THREAD

**Discussion by Cybro0** [Crypto-TII/claasp#61] [Q&A]
Hi, I opted for manual installation which I cloned the git, and next command is to run the "configure. sh" file. Then I noticed that it doesn't have execute permission. So I chmod it and after running the script. I got the error as this: "dependencies_script.sh" permission denied. How to resolve this?. or should I go for docker installation?
![Screenshot 2023-09-01 231910](https://github.com/Crypto-TII/claasp/assets/143835501/33004231-6d1a-4da1-997a-bb7509ed05bf)

**Comment by peacker** [comment]
Hi Cybro0, thanks for your interest in CLAASP! Indeed, the docker image should simplify the installation, and our recommendation is to do that. Note that we are also working on a web application that will allow to access CLAASP with no need of installing anything beside the Chrome browser. We hope to have this available by early 2024. If you really want to go for the full local installation, please let us know and we can try to debug your issue.
At a first sight, is looks that, as a quick solution, running  the installation with sudo might solve the problem.

**Reply by Cybro0** [reply]
Hello, i have a doubt regarding adding a new cipher in CLAASP tool. Can u help me? Any contact?

**Comment by juaninf** [comment]
Hi Cybro0, thanks for your interest in CLAASP,
Could you please send us your question?
thanks
---------------------------------------------------------------------
D.Sc. Juan del Carmen Grados Vásquez
Laboratório Nacional de Computação Científica
Tel: +55 21 97633 3228
(http://www.lncc.br/)
http://juaninf.blogspot.com
---------------------------------------------------------------------

**Reply by Cybro0** [reply]
I have 32 bit value in t3 variable and i have to pass it through S-Box. 
Below code substitutes 8 bits....4 times....

Question is, How do i take this 4 8-bit values as one string or one value in a variable.

![sbox](https://github.com/Crypto-TII/claasp/assets/143835501/2448447d-c379-42ee-bacf-250291e24028)

**Comment by juaninf** [comment]
Hi,
I will suggest to check to use the intermediate output components. You can
see some examples of them in almost of the ciphers that are implemented in
claasp. For example, in Speck.

**Comment by juaninf** [comment]
Have you manage to solve this?

**Comment by juaninf** [comment]
Specifically, in the file speck_block_cipher.py. If it is the last round
also you can use the “cipher output component”. Again you can see several
examples of how to use this in almost of the cipher implemented in claasp.

**Reply by Cybro0** [reply]
Yes, that is for last resultant. But in my case, it is in middle of the round function , as i need to pass the S-Box substituted string to the rotate component. How do i do without the output component

**Comment by juaninf** [comment]
Hi Cybro,
In that case you can use the add_round_output_component component. Since
you need to pass the S-Box substituted string to the rotate component then
you can also directly use the rotate component without using the
add_round_output_component. To do so you need to use the input_id_links
parameter of add_round_output_component. Check the attached image.
[image: Screenshot 2023-12-08 at 3.31.57 AM.png]
---------------------------------------------------------------------
D.Sc. Juan del Carmen Grados Vásquez
Laboratório Nacional de Computação Científica
Tel: +55 21 97633 3228
(http://www.lncc.br/)
http://juaninf.blogspot.com
---------------------------------------------------------------------

**Reply by Cybro0** [reply]
There's no image attached !

**Comment by juaninf** [comment]
Sorry the image is in the body of the email

**Reply by Cybro0** [reply]
I couldn't find the image in body of the email too. Posting that again would be helpful.
My scenario as i pasted my screenshot is, I'm getting last 8 bits of my S_Box output passed to rot component. I have to change my code lines accordingly to get whole S-Box substituted string passed to rot component. How can i do that?
![sbox_to_rot](https://github.com/Crypto-TII/claasp/assets/143835501/de754957-b801-451c-b3ba-14ef6b4508c6)
![sbox_to_rot_code](https://github.com/Crypto-TII/claasp/assets/143835501/052ac84b-8a13-446e-b688-e654c3e8dad4)

**Comment by juaninf** [comment]
Hi Cybro0,
If I understood well you need to do something similar to what is done in
slide 3 of
https://askworkshop.github.io/ask2023/assets/accessory/ASK_slides-juan.pdf.
Please check it and let me know if that help.

**Reply by Cybro0** [reply]
I tried as you done in 7th slide (ToySPN1). But, im getting error of " the length of [input_bit_positions] is not equal to input_bit_size ".
![4sboxes](https://github.com/Crypto-TII/claasp/assets/143835501/aa993e70-bc93-4b22-a803-44dc3515b8e5)

**Comment by juaninf** [comment]
I think that your issue is due the second parameter must be a list of sub
lists and each Sublist must contain the indexes. In your case each sublist
has again a list.

**Reply by Cybro0** [reply]
Thank you, it solved the problem. But i encountered another problem.
While performing evaluate( ) function, im getting this TypeError. Don't know which part of my code is wrong.

![add_output_component_code](https://github.com/Crypto-TII/claasp/assets/143835501/e031232e-62b4-4601-a42d-0ceb2b69d884)
![typeerror](https://github.com/Crypto-TII/claasp/assets/143835501/708a3087-0a1f-4459-a7c2-48389a9dba46)

**Comment by juaninf** [comment]
Good that now it is working. Regarding this second issue. It seems you are
using the constant component. If so, could you please share the line
containing the call tho that method?

---------------------------------------------------------------------
D.Sc. Juan del Carmen Grados Vásquez
Laboratório Nacional de Computação Científica
Tel: +55 21 97633 3228
(http://www.lncc.br/)
http://juaninf.blogspot.com
---------------------------------------------------------------------

**Reply by Cybro0** [reply]
I used constants component in round key function and the resultant round key is passed in round function. Last time i checked, there's no any error in round key function and i got correct output. After i done code for round function (complete code), im getting this error. Can i share you complete code by any means?
 
![1a](https://github.com/Crypto-TII/claasp/assets/143835501/a147ae45-fffc-44d6-81da-227d2fa112c6)
![1b](https://github.com/Crypto-TII/claasp/assets/143835501/e6ab51b5-c168-4ac4-8308-4bbcc62b9709)

**Comment by juaninf** [comment]
Also you can check line 92 of
https://github.com/Crypto-TII/claasp/blob/main/claasp/ciphers/block_ciphers/speck_block_cipher.py


---------------------------------------------------------------------
D.Sc. Juan del Carmen Grados Vásquez
Laboratório Nacional de Computação Científica
Tel: +55 21 97633 3228
(http://www.lncc.br/)
http://juaninf.blogspot.com
---------------------------------------------------------------------

**Reply by Cybro0** [reply]
![speckcomponentanalysis](https://github.com/Crypto-TII/claasp/assets/143835501/ee459821-64cb-41a6-9f2b-eff12bd179d2)
How to see the plotted graph of component analysis?

**Comment by juaninf** [comment]
In at least one of your calls to add constant component you’re not  passing
the parameters correctly. Check them again. It seems is something similar
to your previous issue.

---------------------------------------------------------------------
D.Sc. Juan del Carmen Grados Vásquez
Laboratório Nacional de Computação Científica
Tel: +55 21 97633 3228
(http://www.lncc.br/)
http://juaninf.blogspot.com
---------------------------------------------------------------------

**Reply by Cybro0** [reply]
How can i generate that report to parse for statistical tests?

**Reply by Cybro0** [reply]
![daataseterror](https://github.com/Crypto-TII/claasp/assets/143835501/e5ed0f24-e46c-4271-af4f-ae8e082d8a5f)
Im getting this error for dataset generation

**Comment by juaninf** [comment]
Hi you can use savefigs
https://chartio.com/resources/tutorials/how-to-save-a-plot-to-a-file-using-matplotlib/
---------------------------------------------------------------------
D.Sc. Juan del Carmen Grados Vásquez
Laboratório Nacional de Computação Científica
Tel: +55 21 97633 3228
(http://www.lncc.br/)
http://juaninf.blogspot.com
---------------------------------------------------------------------

**Comment by Adelaide-emporor** [comment]
Hello,

I've encountered an issue during the installation process. Could you please advise on how to resolve it? I've followed the steps to execute make local-installation strictly on Ubuntu 20.04. What should I do next to address this problem and successfully install the Claasp tool? I'm eagerly awaiting your response.

Thank you.

![屏幕截图(83)](https://github.com/Crypto-TII/claasp/assets/87794783/1f3dee7b-6cba-4094-9cfa-6d9918fece40)

**Reply by juaninf** [reply]
Hi @Adelaide-emporor, I apologize for the delayed response. Thank you for bringing this to our attention. We are currently analyzing the issue and will provide a more detailed answer soon.
